File: puzzles-leet-code/src/main/java/org/krmdemo/techlabs/interview/meta/FB_Prep_Greedy__Element_Swapping.java
package org.krmdemo.techlabs.interview.meta;

import java.util.*;

/**
 * <h3><a href="https://www.metacareers.com/profile/coding_practice_question/?problem_id=838749853303393&c=3513648482275061&psid=275492097255885&practice_plan=0&b=0111122">
 *     Greedy #2. Element Swapping.
 * </a></h3>
 * Given a sequence of <code>arr.length</code> integers <b><code>arr</code></b>,
 * determine the <a href="https://en.wikipedia.org/wiki/Lexicographic_order">lexicographically smallest</a> sequence
 * which may be obtained from it after performing at most <b><code>K</code></b> element swaps,
 * each involving a pair of <u>consecutive elements</u> in the sequence.
 * <hr/>
 * <small>
 * <u><i>Note:</i></u> A <code>list x</code> is lexicographically smaller
 * than a different equal-length <code>list y</code> if and only if, for the earliest index  at which the two lists differ,
 * <code>x</code>'s element at that index is smaller than <code>y</code>'s element at that index.
 * </small>
 * <h5>Constraints:</h5><pre>
 * 1 ≤ arr.length ≤ 10^3
 *      1 ≤ K ≤ 1000
 *     1 ≤ arr[i] ≤ 10^6
 * </pre>
 *
 * @see org.krmdemo.techlabs.leet_code_0000_1000.Problem_031__Next_Permutation (more advanced problem)
 */
public interface FB_Prep_Greedy__Element_Swapping {

    /**
     * Solution entry-point.
     *
     * @param arr an array of positive integers (values don't really matter) of size not more than <code>1000</code>
     * @param K the exact number swaps to perform for <u>only consecutive pair</u> of elements
     * @return the lexicographically smallest array after those <code>K</code> swaps
     */
    int[] findMinArray(int[] arr, int K);

    /**
     * The proposed solution is based on {@link PriorityQueue}
     * and appears to be a classic greedy algorithm (the strict proof is not trivial)
     */
    enum Solution implements FB_Prep_Greedy__Element_Swapping {
        IN_PLACE;

        @Override
        public int[] findMinArray(int[] arr, int K) {
            System.out.println("arr --> " + Arrays.toString(arr));
            System.out.println("initially K = " + K);
            for (int index = 0; index < arr.length; index++) {
                if (K <= 0) {
                    break;
                }
                OptionalInt minIndexOpt = nextIndexOfMin(arr, K, index);
                if (minIndexOpt.isPresent()) {
                    int minIndex = minIndexOpt.getAsInt();
                    System.out.printf("- swapping [%d]%d and [%d]%d K: %d - %d = %d; ",
                        index, arr[index], minIndex, arr[minIndex],
                        K, minIndex - index, K - (minIndex - index));
//                    swapElems(arr, index, minIndex);  // <-- this is not correct
                    bubbleValue(arr, index, minIndex);
                    K -= (minIndex - index);
                    System.out.println("--> " + Arrays.toString(arr));
                };
            }
            System.out.println("finally K = " + K);
            System.out.println("finally arr --> " + Arrays.toString(arr));
            if (Arrays.equals(arr, Arrays.stream(arr).sorted().toArray())) {
                System.out.println("<<< the array is sorted !!! >>>\n");
            } else {
                System.out.println(">>> the array is still un-sorted, but no swaps remain <<<\n");
            }
            return arr;
        }

        /**
         * Looking for the minimal element among not more than next <code>K</code> elements
         * since the element with the passed index <code>indexBefore</code>.
         *
         * @param arr an array to search the index of minimal element at
         * @param K the maximum number of next elements to search the minimal one
         * @param indexBefore the index of element whose value must be <b>greater</b> than next minimal one
         * @return the optional wrapper over the found index of the found next minimal element
         */
        OptionalInt nextIndexOfMin(int[] arr, int K, int indexBefore) {
            int minValue = arr[indexBefore];
            OptionalInt minIndex = OptionalInt.empty();
            int rightBound = Math.min(indexBefore + K + 1, arr.length);
            for (int i = indexBefore + 1; i < rightBound; i++) {
                if (minValue > arr[i]) {
                    minValue = arr[i];
                    minIndex = OptionalInt.of(i);
                }
            }
            return minIndex;
        }

        /**
         * Bubble the element to the target index by consecutive swaps
         *
         * @param arr an array to swap elements
         * @param targetIndex the raget index of element that is going to bubble
         * @param indexToBubble the initial index of element to bubble
         */
        private void bubbleValue(int[] arr, int targetIndex, int indexToBubble) {
            for (int index = indexToBubble; index > targetIndex; index--) {
                int bubbleValue = arr[index];
                arr[index] = arr[index - 1];
                arr[index - 1] = bubbleValue;
            }
        }

        /**
         * Swapping the elements in array - this MUST NOT be used in this puzzle
         *
         * @param arr an array to swap elements
         * @param indexOne the first index to swap
         * @param indexTwo the second index to swap
         */
        private void swapElems(int[] arr, int indexOne, int indexTwo) {
            int valueOne = arr[indexOne];
            arr[indexOne] = arr[indexTwo];
            arr[indexTwo] = valueOne;
        }
    }
}

File: puzzles-leet-code/src/test/resources/org/krmdemo/techlabs/interview/meta/FB_Prep_Greedy__Element_Swapping.py
"""
`META-Prep - Greedy #2. Element Swapping`_

To solve the problem of finding the lexicographically smallest sequence after at most k element swaps
(where each swap involves a pair of consecutive elements), we can follow a greedy approach.

Approach:

    Greedy Strategy:
    At each step, we try to place the smallest possible number in the current position,
    using the allowed swaps. If we are at index i, we look for the smallest element
    within the next k elements (since we can swap only consecutive elements) and move it to the current position.
    This reduces the problem to progressively fixing the sequence from left to right while respecting the swap limit k.

    Choosing the Smallest Element:
    - For each position i, we need to find the smallest element in the subarray arr[i, i+1, ..., i+k]
      (as we can swap only up to k positions ahead).
    - We swap this smallest element to position i, and each swap will decrement k by the number of swaps used.

    Termination: We continue until either:
    - The array is sorted lexicographically.
    - Or we've made k swaps.

.. _`META-Prep - Greedy #2. Element Swapping`: https://www.metacareers.com/profile/coding_practice_question/?problem_id=838749853303393&c=3513648482275061&psid=275492097255885&practice_plan=0&b=0111122
"""

def lexicographically_smallest_sequence(arr: list[int], k: int) -> list[int]:
    """
    Solution entry-point

    :param arr: initial array
    :param k: the maximum number of swaps
    :return: the result array
    """

    n = len(arr)

    # Iterate through the array
    for i in range(n):
        if k <= 0:
            break  # No swaps left, stop

        # Find the smallest element within the next k elements
        min_idx = i
        for j in range(i + 1, min(i + k + 1, n)):
            if arr[j] < arr[min_idx]:
                min_idx = j

        # If the minimum element is not at the current position, we swap it to the front
        if min_idx != i:
            # Swap elements from min_idx to i, which will bring the min element to position i
            for j in range(min_idx, i, -1):
                arr[j], arr[j - 1] = arr[j - 1], arr[j]
            k -= (min_idx - i)  # Decrease the number of available swaps

    return arr


# Example usage:
for k in range(1, 10):
    arr = [5, 2, 9, 1, 5, 6]
    print(f"k = {k}; {arr = } --> ", end='')
    lexicographically_smallest_sequence(arr, k)
    print(f"result --> {arr}", end='')
    if sorted(arr) == arr:
        print(' SORTED !!!')
    else:
        print(' still NOT sorted')

# Output must look something like:
# >>> k = 1; arr = [5, 2, 9, 1, 5, 6] --> result --> [2, 5, 9, 1, 5, 6] still NOTsorted
# >>> k = 2; arr = [5, 2, 9, 1, 5, 6] --> result --> [2, 5, 1, 9, 5, 6] still NOT sorted
# >>> k = 3; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 5, 2, 9, 5, 6] still NOT sorted
# >>> k = 4; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 9, 5, 6] still NOT sorted
# >>> k = 5; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 5, 9, 6] still NOT sorted
# >>> k = 6; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 5, 6, 9] SORTED !!!
# >>> k = 7; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 5, 6, 9] SORTED !!!
# >>> k = 8; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 5, 6, 9] SORTED !!!
# >>> k = 9; arr = [5, 2, 9, 1, 5, 6] --> result --> [1, 2, 5, 5, 6, 9] SORTED !!!
